### DISCUSSION TITLE
Python for Android - Executable

### DISCUSSION THREAD

**Discussion by eakteam** [termux/termux-packages#8366] [Q&A] [answered]
Hello, i am having difficulties understanding why executing python module is not working.
So the steps are like that : 

Edited `TERMUX_APP_PACKAGE="com.somepackage"` at termux-packages/scripts/properties.sh
Edited termux-packages/scripts/build/termux-step-setup-variables.sh and add thoose 2 lines : 
`: "${TERMUX_PREFIX:="/data/data/com.somepackage/files/usr"}"`
`: "${TERMUX_ANDROID_HOME:="/data/data/com.somepackage/files/home"}"`

Build python -> ./build-package.sh python

After it is done, get all the folders produced inside .deb file and upload to Android `/data/local/tmp` because it have execute permission allowed from adb shell to allow testing.

When trying to execute it with `./python3`
It fails with : `CANNOT LINK EXECUTABLE "./python3": library "libandroid-support.so" not found: needed by main executable`

But i am sure that this library is there and existing.

**Comment by ghost** [comment, marked as answer]
That's because packages aren't standalone. They have dependencies, e.g. libraries such as `libc.so`, `libandroid-support.so` and many others. All files must be placed at specific locations which also make packages not relocatable.

Say you have compiled package with prefix `/data/data/com.somepackage/files/usr`. That means:
* Package installation root would be `/data/data/com.somepackage/files/usr`.
* Data files will be looked at `/data/data/com.somepackage/files/usr/share`.
* Libraries will be looked at `/data/data/com.somepackage/files/usr/lib`.
* Temporary files will be stored at `/data/data/com.somepackage/files/usr/tmp`.

But you are placing everything into `/data/local/tmp`. That would not work at least because compiled-in library search path is different. In case of `python` package you will additional issue where `python` will be unable to locate its stdlib.

**Reply by eakteam** [reply]
I understand that but i have also copied python folder to that `/data/data/com.somepackage/files/usr/lib` too... So it should work.

**Reply by ghost** [reply]
`/data/data/com.somepackage/files/usr` should contain all dependency packages for `python` package:
* ca-certificates
* gdbm
* libandroid-support
* libbz2
* libcrypt
* libexpat
* libffi
* liblzma
* libsqlite
* ncurses
* openssl
* readline
* zlib

and be readable/executable from ADB. Normally ADB doesn't have access to application directories.

**Reply by eakteam** [reply]
The second screen is root folder and the first is lib folder... Where i am wrong ?
![lib_root](https://user-images.githubusercontent.com/8820741/147422802-421b7469-fed9-4d5c-9752-0ec0f9420de3.png)
![root](https://user-images.githubusercontent.com/8820741/147422803-00a846ff-72af-4b62-bd60-fcd79b45448f.png)

**Reply by eakteam** [reply]
And what if i change `/data/data/com.somepackage/files/usr` to `/data/local/tmp` ? it should work ?

**Reply by eakteam** [reply]
Can You help please ?

**Reply by Yonle** [reply]
`/data/local/tmp` is unread & unwriteable on android. The path is correct. You just didn't put the required library on your `/data/data/com.somepackage/files/usr/lib`, Plus you're testing on `/data/local/tmp` which is not gonna work.

**Reply by ghost** [reply]
@eakteam You put only python files in lib directory. Iny previous comments I wrote that you need to put files of **every python package dependency** and in the last comment I provided a list of dependency packages.

Packages are not standalone. One package depend on other(s) and form dependency chain. So get all Deb files required to run Python and extract them to prefix directory.

If you are going to use ADB, your prefix should be `/data/local/tmp` and therefore lib path is going to be `/data/local/tmp/lib`. As paths are hardcoded into binaries, changing prefix will require rebuilding.
